### DISCUSSION TITLE
What is the best practice for mocking nested functions for testing?

### DISCUSSION THREAD

**Discussion by mhkafadar** [asomers/mockall#384] [Questions] [answered]
I will simplify my use case and implementation to make my question brief.

There are 3 functions in my lib.rs

fetch_words_from_database() -> Vec<&'a str>
fetch_words_from_api() -> Vec<&'a str>
find_intersection() -> Vec<&'a str>
I want to unit test find_intersection function so I have to mock functions 1 and 2. How to achieve that?

```rust
fn find_intersection() -> Vec<&'a str> {
        let words_database = fetch_words_from_database();
        let words_api = fetch_words_from_api();
        let words_api: HashSet<_> = words_api.into_iter().collect();
        
        let mut intersection: Vec<&'a str> = Vec::new();
        
        for word in words_database {
            if words_api.contains(word) {
                intersection.push(word)
            }
        }
        
        intersection
}
```

I am looking for something like jest.spyOn.

**Comment by asomers** [comment, marked as answer]
I don't know anything about typescript, but this seems like a simple question.  Just mock those two functions according to the docs at https://docs.rs/mockall/latest/mockall/#static-methods or https://docs.rs/mockall/latest/mockall/#modules .

**Reply by mhkafadar** [reply]
Yes thanks. But what about not using Trait. Is it possible to mock a function which not defined in a trait? Because those functions will not be used by multiple structs, they will be just used in once in whole project. So I don't want to define them as a trait just for testing/mocking purposes.
Thanks!

**Reply by asomers** [reply]
Yes, certainly.  Just look at the second example I linked: it doesn't use a trait.  There is also https://docs.rs/mockall/latest/mockall/#mocking-structs , which works with static methods too.
